core/src/main/scala/org/http4s/blaze/channel/nio1/NIO1SocketServerGroup.scala

```diff
@@ -49,30 +54,117 @@ object NIO1SocketServerGroup {
 final class NIO1SocketServerGroup private (pool: SelectorLoopPool, channelOptions: ChannelOptions)
     extends ServerChannelGroup {
 
-  @volatile
-  private[this] var isClosed = false
   private[this] val logger = getLogger
-  private[this] val selector = Selector.open()
-  private val t = new AcceptThread()
+  // Also acts as our intrinsic lock.
+  private[this] val listeningSet = new mutable.HashSet[ServerChannelImpl]()
+  // protected by synchronization on the intrinsic lock.
+  private[this] var isClosed = false
+
+  // Closing delegates to the `ServerChannelImpl` which
+  // ensures only-once behavior and attempts to close the
+  // channel within the `SelectorLoop`, if it's still running.
+  // That should help alleviate many of the exceptions associated
+  // with performing operations on closed channels, etc.
+  private[this] class SocketAcceptor(
+      key: SelectionKey,
+      ch: ServerChannelImpl,
+      service: BufferPipelineBuilder)
+      extends Selectable {
+
+    // Save it since once the channel is closed, we're in trouble.
+    private[this] val closed = new AtomicBoolean(false)
+
+    override def opsReady(unused: ByteBuffer): Unit =
+      if (key.isAcceptable) {
+        try acceptNewConnections()
+        catch {
+          case ex: IOException =>
+            closeWithError(ex)
+        }
+      }
+
+    override def close(): Unit =
+      if (closed.compareAndSet(false, true)) {
+        logger.info(s"Listening socket(${ch.socketAddress})")
```

Comment:
This statement confuses me.

Reply:
It's clearly missing something... I'll fix it.